src/main.cpp

```diff
@@ -669,9 +669,15 @@ Args parse_args(int argc, char* argv[])
     if (optind == argc) {
       args.search = "*:*";
     } else if (optind == argc - 1) {
-      args.search = argv[optind];
-      if (args.search == "*") {
+      std::string_view val(argv[optind]);
+      if (val == "*") {
         args.search = "*:*";
+      } else if (val.find_first_of(":*") != std::string::npos) {
+        args.search = val;
+      } else if (std_filesystem::exists(val)) {
+        args.filename = val;
+      } else {
+        args.search = val;
```

Comment:
```suggestion
      } else {
        args.filename = val;
```

I was thinking we could just treat everything not containing `:*` as a file - then error messages would still look sensible for non-existent files. All probes (other than BEGIN and END) will contain one of the `:*` characters so error messages for anything resembling them would look good too

Reply:
I do find it a bit odd that search doesn't work for substrings in the listing path e.g. this message is not great
```
$sudo bpftrace -l tracepoint
stdin:1:1-11: ERROR: tracepoint probe type requires 2 arguments, found 0
tracepoint
~~~~~~~~~~
```

Reply:
this code breaks existing test: `-bpftrace l '\n'`. It's registered under "errors on invalid character in search expression" name

Reply:
Looking at why that test was added (#666), it was for some regex matching code that couldn't handle some characters. That regex matching code was completely removed in #1549. The lexer now detects invalid characters so invalid character-detection is a parser thing now, not specific to probe listing.

I think this runtime test is unnecessary and can be deleted.

If you fancy it, you could try replacing it with a Parser unit test instead, attempting to parse an invalid character like:
```
test_parse_failure("\\", R"(
```

Here's an example of a parser test:
https://github.com/bpftrace/bpftrace/blob/8ac0a49fe9481221260147aa722b8e150f8560e6/tests/parser.cpp#L1113-L1119)